### PR TITLE
client: optimize the mechanism of creating TSO dispatcher

### PR DESCRIPTION
**Pull request by JmPotato**
Signed-off-by: JmPotato <ghzpotato@gmail.com>

### What problem does this PR solve?

Due to `client.tsoDispatcher` is updated with the `baseClient.allocators`, sometimes they are not consistent because the updating is out of sync. This PR fixes this problem.

### What is changed and how it works?

* Update `client.tsoDispatcher` more reasonably.
* Add GC for the unused TSO allocator and dispatcher.
* Add some logs.

### Check List

Tests

- Unit test
- Integration test

### CODE REVIEW

File: client/base_client.go [hunk 1]
```diff
@@ -265,6 +275,7 @@ func (c *baseClient) updateLeader() error {
 		c.updateURLs(members.GetMembers())
 		return c.switchLeader(members.GetLeader().GetClientUrls())
 	}
+	c.scheduleCheckTSODispatcher()
```
Comment (rleungx) [inline client/base_client.go:278]: Does it need to be moved into the loop?

File: client/client.go [hunk 2]
```diff
@@ -311,41 +349,35 @@ func (c *client) tsLoop() {
...
-				// Each goroutine is responsible for handling the tso stream request for its dc-location.
-				// The only case that will make the dispatcher goroutine exit
-				// is that the loopCtx is done, otherwise there is no circumstance
-				// this goroutine should exit.
-				go c.handleDispatcher(loopCtx, dcLocation, dispatcher.(chan *tsoRequest))
-			}
-			return true
-		})
 		select {
+		case <-c.checkTSODispatcherCh:
```
Comment (Yisaer) [inline client/client.go:353]: Why we do nothing here?

**Review by Yisaer — COMMENTED** [review]
Rest LGTM

**Review by Yisaer — APPROVED** [review]
LGTM

File: client/base_client.go [hunk 3]
```diff
@@ -41,7 +41,8 @@ type baseClient struct {
 	// dc-location -> TSO allocator leader URL
 	allocators sync.Map // Store as map[string]string
 
-	checkLeaderCh chan struct{}
+	checkLeaderCh        chan struct{}
+	checkTSODispatcherCh chan struct{}
```
Comment (rleungx) [inline client/base_client.go:45]: It seems this channel is not initialized?

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**